File: app/src/main/java/com/yalantis/data/source/repository/RepositoryLocalDataSource.java
package com.yalantis.data.source.repository;

import android.support.annotation.NonNull;

import com.yalantis.data.Repository;
import com.yalantis.data.RepositoryFields;
import com.yalantis.data.source.base.BaseLocalDataSource;

import java.util.List;

import io.realm.Realm;
import io.realm.Sort;
import rx.Observable;
import rx.Single;

/**
 * Created by irinagalata on 12/1/16.
 */

class RepositoryLocalDataSource extends BaseLocalDataSource implements RepositoryDataSource {

    @Override
    public Single<List<Repository>> getRepositories(@NonNull String organization) {
        return Observable.just((List<Repository>) getRealm()
                .where(Repository.class)
                .findAllSorted(RepositoryFields.STARS_COUNT, Sort.DESCENDING))
                .toSingle();
    }

    @Override
    public void saveRepositories(final List<Repository> repositories) {
        getRealm().executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                realm.copyToRealmOrUpdate(repositories);
            }
        });
    }

    @Override
    public void clearRepositories() {
        getRealm().executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                realm.delete(Repository.class);
            }
        });
    }

    @Override
    public boolean isEmpty() {
        return getRealm().where(Repository.class).count() > 0;
    }

}

File: app/src/main/java/com/yalantis/data/source/repository/RepositoryLocalDataSource.kt
package com.yalantis.data.source.repository

import com.yalantis.data.Repository
import com.yalantis.data.RepositoryFields
import com.yalantis.data.source.base.BaseLocalDataSource
import io.reactivex.Single
import io.realm.Sort

/**
 * Created by irinagalata on 12/1/16.
 */

internal class RepositoryLocalDataSource : BaseLocalDataSource(), RepositoryDataSource {

    override fun getRepositories(organization: String): Single<List<Repository>> =
            Single.just(realm.where<Repository>(Repository::class.java)
                .findAllSorted(RepositoryFields.STARS_COUNT, Sort.DESCENDING) as List<Repository>)

    override fun saveRepositories(repositories: List<Repository>) {
        realm.executeTransaction { realm -> realm.copyToRealmOrUpdate(repositories) }
    }

    override fun clearRepositories() {
        realm.executeTransaction { realm -> realm.delete(Repository::class.java) }
    }

    override fun isEmpty(): Boolean = realm.where<Repository>(Repository::class.java).count() > 0

}
